### DISCUSSION TITLE
Add a react virtual scrolling library

### DISCUSSION THREAD

**Discussion by maxpatiiuk** [specify/specify7#5597] [Feature Requests]
Need to add a react virtual scrolling library to improve rendering performance.

Can be used in the following places:
- Autocomplete results
- Preparations list
- Query results
- FormTable

Probably there are many other places where this would be applicable.
Though, in some places explicit pagination would make more sense. See https://github.com/specify/specify7/issues/1398

**Comment by maxpatiiuk** [comment]
@CarolineDenis did research on this and faced some obstacles. Feel free to write down the things you found before we forget for future reference.

A summary from what I can remember:
Virtual scrolling libraries are still not very satisfactory - they add too many constraints:
 - They require hard-coding item height
 - They don't handle items of varying height very well (this happens often due to overflow)
 - They require hard-coding container height
 - They require specific containers (`<div>` instead of `<tr>` - thus making this almost unusable for tables)
 - They don't handle container resize well

Thus, for now a better solutions are (depending on which is more applicable):
- Do infinite scrolling (like in Query Results or Attachments viewer), where more results are fetched automatically as soon as user scrolls close to the end - quite bug-prone (there been a lot bugs related to this in query builder and attachments); performance degrates as more records are loaded
- Pagination (#1398) - simple but adds more friction

**Comment by CarolineDenis** [comment]
Paginator component created in this PR: https://github.com/specify/specify7/issues/1398

**Comment by grantfitzsimmons** [comment]
Fixed by #3195

**Comment by maxpatiiuk** [comment]
#3195 addresses that for record sets and query results dialogs only.

1. There are other dialogs that haven't been addressed yet (data sets, reports)
2. There are places that need virtual scrolling library still:
    - Autocomplete results
    - Preparations list
    - Query results
    - FormTable

**Comment by maxpatiiuk** [comment]
i.e, when CSIRO was complaining about poor performance for large pick lists, or when @grantfitzsimmons was complaining about poor autocomplete performance when doing search with a single character - both of those could be partially solved by adding virtual scrolling library or a similar solution
